### PR TITLE
Allow to load python object from yaml

### PR DESCRIPTION
**Pull request by sileht**
It can be interesting to write custom object to compare values.

For example, I need to ensure an output is equal to .NAN

Because .NAN == .NAN always returns false. We currently can't compare it
with assert_equals().

With the unsafe yaml loader we can register a custom method to check
NAN, for example:

    class IsNAN(object):
        @classmethod
        def constructor(cls, loader, node):
            return cls()

        def __eq__(self, other):
            return numpy.isnan(other)

    yaml.add_constructor(u'!ISNAN', ISNAN.constructor)

### CODE REVIEW

**Comment by cdent** [comment]
Seems reasonable but I'm having a hard time visualizing. Have you got an example where you're already using it?

I suppose since we're in a controlled environment there's no reason to insisnt on safe_load.

**Comment by sileht** [comment]
I plan to use this for: https://github.com/gnocchixyz/gnocchi/pull/446/files#diff-f12b9f5f43a3ce0d98cc7e1dac0cee7fR432

**Comment by sileht** [comment]
I plan to replace .NAN by !ISNAN to make the  test pass

**Comment by cdent** [comment]
Ah, okay. Thanks that makes it a lot more clear. If you could add a section to the faq.rst about how to make use of this that would be great, but if you don't get to it before I get to merging it (sometime tomorrow, going to bed now) I'll add something. Thanks.

**Comment by FND** [comment]
I understand the idea - it's a good one which might open up interesting new avenues - but I'm pretty uncomfortable with this assumption:

> I suppose since we're in a controlled environment there's no reason to insisnt on safe_load.

I'd much rather individual suites opt into that.

**Comment by cdent** [comment]
> I'd much rather individual suites opt into that.

So your preference would be that the test runner set a flag or something of that ilk which would switch away from using `safe_load`? That might be useful, but it's probably also worthwhile to bone up on the difference between `safe_load` and `load` (I haven't yet because of time).

**Comment by cdent** [comment]
For reference, some conversations with FND about this starts at: http://p.anticdent.org/1AsR

We have four options for moving forward:

1. merge this as is
0. add a flag to build_tests to switch in load for safe_load upon request
0. environ var that does same
0. reject this

4 is not a good option.

Will come back to this asap.

**Comment by sileht** [comment]
I prefer 1. and 2.

**Comment by FND** [comment]
I guess 2. (plus gabbi-run support) would (more or less) address my (still fairly vague and unarticulated) concerns - though I'm not actually sure about that. ¯\_(ツ)_/¯ Sorry I can't be more helpful at this time.

**Comment by sileht** [comment]
I'm OK with 2, I just need a name for the new parameter. safe_yaml_loading ? unsafe_yaml_loading ? unsafe_yaml ? allow_custom_yaml_symbol ? enable_custom_symbol ?

**Comment by cdent** [comment]
How about `safe_yaml`, defaults to `True`, and callers can set it to`False` if they'd like otherwise?

**Comment by FND** [comment]
After (yet another) [lengthy IRC discussion](http://p.anticdent.org/2tXU) about my concerns (mostly about separation of concerns: dict-style data model vs. YAML representation), cdent has coerced my blessing (not that that was mine to give; the best I could hope for was more filibustering).

So as far as I'm concerned, `build_tests(…, safe_yaml=False)` seems fine. (I briefly tried to find a more salient term in places like [tenderlove's write-up](http://tenderlovemaking.com/2013/02/06/yaml-f7u12.html), but failed to come up with anything more alarming.)

**Comment by sileht** [comment]
Pull-request updated, HEAD is now d48213c7ffe2a03d37db2e1b6698736b4b48d265

File: gabbi/runner.py [hunk 1]
```diff
@@ -224,6 +227,14 @@ def _make_argparser():
         choices=['all', 'body', 'headers'],
         help='Turn on test verbosity for all tests run in this session.'
     )
+    parser.add_argument(
+        '--unsafe-yaml',
+        dest='safe_yaml',
+        action='store_false',
+        default=True,
+        help='Trun on recognization of Python object in additional of '
```
Comment (cdent) [inline gabbi/runner.py]: Turn on recognition of Python objects in addition to standard YAML tags

**Comment by sileht** [comment]
Pull-request updated, HEAD is now e0114ea4c8d3bde5988facfcfbc5d0601365761c

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 3fcf2ce3d27da13af5414ffed13858d1fbdcc8dd

File: gabbi/runner.py (hunk 1 above)
Comment (sileht) [reply gabbi/runner.py]: done

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 98adca65e05b7de4f1ab2bf90ab0521af5030f35

**Pull request merged**